lib/Doctrine/DBAL/Platforms/DB2Platform.php

```diff
@@ -899,4 +899,13 @@ protected function getReservedKeywordsClass()
     {
         return Keywords\DB2Keywords::class;
     }
+
+    public function getListTableCommentsSQL(string $table) : string
```

Comment:
> We select most of the table details in a single SQL.

Actually, there is one query to fetch all table names, and then, per table, there is one query for all the columns, one for all the indexes and one for the foreign keys.

https://github.com/doctrine/dbal/blob/master/lib/Doctrine/DBAL/Schema/AbstractSchemaManager.php#L245-L277

But the first query used to list the table names returns only the names.

When there are additional table metadata to be fetched (as in the MySQLSchemaManager), the table comments are fetched in the same query as the other metadata: https://github.com/doctrine/dbal/blob/master/lib/Doctrine/DBAL/Platforms/MySqlPlatform.php#L383-L395
You can notice that in the current implementation, there is already N+1 queries to fetch MySQL metadata (including table comments).

We may change `listTableNames` into something like `listTableNamesAndComments` or `listTableNamesAndMetadata` that would return the table names AND the comments at once. What do you think?

Reply:
Please use permalinks when referring to code fragments. This way it's guaranteed that the same fragment will be referred regardless of the future code changes.

Reply:
Some platforms like Oracle build schema introspection views on the fly and it's very inefficient to build them N times instead of just once. We should avoid introducing an API which will knowingly have performance problems.

I'd say a comment is part of the metadata/details by its meaning. API-wise, why would we want to have a separate method which returns only table names and their comments?

Would it make sense to introduce something like `AbstractPlatform::getListTablesWithDetails()`? Then `AbstractSchemaManager::listTables()`, instead of getting a list of tables and fetching their details one by one would call the new method.

Reply:
> Some platforms like Oracle build schema introspection views on the fly and it's very inefficient to build them N times instead of just once. We should avoid introducing an API which will knowingly have performance problems.
>
> I'd say a comment is part of the metadata/details by its meaning. API-wise, why would we want to have a separate method which returns only table names and their comments?

I wholeheartedly agree.

> Would it make sense to introduce something like `AbstractPlatform::getListTablesWithDetails()`? Then `AbstractSchemaManager::listTables()`, instead of getting a list of tables and fetching their details one by one would call the new method.

I'll give this a try.

Reply:
@morozov I started working on this, but I ran into an issue.

Currently, the `listTables` method calls `listTablesDetails` in a loop. This is logical because we need to fetch all columns of all tables.

https://github.com/doctrine/dbal/blob/29486f06cae54166f19423b6bf87bab73fb58ec1/lib/Doctrine/DBAL/Schema/AbstractSchemaManager.php#L245-L260

The instantiation of the `Table` object is done in `listTablesDetails`:

https://github.com/doctrine/dbal/blob/29486f06cae54166f19423b6bf87bab73fb58ec1/lib/Doctrine/DBAL/Schema/AbstractSchemaManager.php#L267-L277

In my current attempt, `listTables` looks like this (in pseudo-code):

```
function listTables() {
    $tableRows = platform->getTableListWithMetadata()

    foreach ($tableRows as $tableRow) {
        $table = $this->listTableDetails($tableRow['name']);

        $table->setComment($tableRow['comment'])
        $tables[] = $table;
    }
    return $tables;
}
```

This works perfectly well... except that if a user calls `listTablesDetails` directly (instead of going through `listTables`, the table object returned will not contain the metadata.

This means I should consider writing 2 variants of `listTablesDetails`. One that can be called directly and that actively fetches the metadata. And one that is provided the metadata.
I can certainly do that is you wish, but it adds some complexity to the code. Do you want me to keep working in that direction or is the additional complexity not worth performance benefit?

Reply:
Yeah… effort-wise, this is getting out of control. This is solvable in the way that there should be a method which returns all columns of all tables so it gets called only once for all tables but it's really out of scope.

Let's leave it as is and thank you for giving it a try.